### ISSUE TITLE
12$: Warn when shadowing variable with declaration only

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#377]
Example:
```js
    let idlSourceOutputStream;
    if (typeof idlSourcePath !== 'undefined') {
        // Intended: idlSourceOutputStream = fs.createWriteStream(idlSourcePath);
        let idlSourceOutputStream = fs.createWriteStream(idlSourcePath);
    } else {
        // Intended: idlSourceOutputStream = fs.createWriteStream("idls");
        let idlSourceOutputStream = fs.createWriteStream("idls");
    }
```

If a `let`-declared variable is shadowed and the shadowing variable is never referenced (only declared), the programmer probably meant to assign to the shadowed variable instead of declaring a new variable. We should report a warning for such code.

**Comment by strager** [comment]
We don't want to report a diagnostic in the following cases:
```js
let x;
function f() {
  let x = 42; // different function
}
```
```js
let x;
if (true) {
  let x; // no initializer
}
```
```js
var x;
if (true) {
  var x = 42;  // var is weird
}
```

**Comment by strager** [comment]
another example (unsure if this should diagnose or not):
```js
for (let x of xs) {
  let x = 3;
}
```

**Comment by strager** [comment]
Shouldn't diagnose:
```js
let x = 3;
for (let x of xs) {
}
```

**Comment by strager** [comment]
Implemented in commit 8ab66c77fe24470e4e4040d4172ea6bfefad4553.
